packages/component-library/components/Heading/Heading.tsx

```diff
@@ -25,18 +26,28 @@ export type AllowedTags =
   | "h5"
   | "h6"
 
+export type AllowedColors =
+  | "dark"
+  | "dark-reduced-opacity"
+  | "white"
+  | "white-reduced-opacity"
+  | "positive"
+  | "negative"
+
 export interface HeadingProps {
   classNameAndIHaveSpokenToDST?: string
   children: React.ReactNode
   tag?: AllowedTags
   variant: HeadingVariants
+  textColor?: AllowedColors
 }
 
 export const Heading = ({
   classNameAndIHaveSpokenToDST,
   children,
   tag,
   variant,
+  textColor = "dark",
```

Comment:
Have you thought about what this might mean for other components that accept colour? Will we end up with `textColor`, `backgroundColor`, `iconColor`, etc?

Reply:
Yes and yes – that's the intention. Do you see any problems arising from that? 
This came up in a meeting I had with Marc and Di where the prop used to be called `color`, but they were pretty keen on it being `textColor` to differentiate it from future stuff.

Reply:
> Do you see any problems arising from that?
Yes. Following this will result in name collisions with CSS properties (`background-color` is `backgroundColor` when represented as an object in React). Will this make it confusing for developers? Will they (incorrectly) think that `backgroundColor` should take hex values? 

My 2c is we should always avoid names that could overload the meaning of our props, where possible. Foreground/background is my preference for this reason.

Reply:
It's also slightly more succinct - 2 color props to remember instead of many (.*Color could result in SO MANY combinations).

Reply:
I like it that if you have an icon with text, you could just specify `foregroundColor` rather than having to specify `textColor` and `iconColor`. It's less flexible if you did actually want a different text color to icon colour, but that seems like an edge case we shouldn't have to worry about just yet?

Reply:
`backgroundColor` is a tricky one - I can't imagine what name you could come up with that wasn't the same as the `background-color`. But the TS types should be enough to make it clear that it isn't a hex value? (Would it even compile if you gave it a hex value?). I don't think it makes sense for headings to have background colors, so we don't have to worry about background color just yet.

Reply:
I'm thinking to drop the `color` from the name with the `foreground/background` suggestion. In your edge case, that could be solved if we lean more heavily on composition rather than overloading components with props. As an example:
```
<Tab icon="help" iconColor="dark" backgroundColor="light" textColor="wisteria-800">
   Example
</Tab>
```

Better utilisation of composition:
```
<Tab fg="wisteria-800" bg="light">
  <Icon variant="help" fg="dark"/>
  Example
</Tab>
```

Reply:
@ActuallyACat That last example you posted resolves the only qualm I had with the fg/bg idea (what if the icon is a different colour to the text?). I'll go with fg/bg, although I wonder if it could be confusing to a developer if they find a component with a type `fg: string`, and there's no `bg`, which will be the case here. I can imagine some people going "wtf is `fg`?" For that reason I'm leaning towards `foregroundColor`, even if it's clumsier.

Reply:
Maybe `fgColor` (and `bgColor`)?

Reply:
> Yes. Following this will result in name collisions with CSS properties (background-color is backgroundColor when represented as an object in React). Will this make it confusing for developers? Will they (incorrectly) think that backgroundColor should take hex values?

I just had a thought – wouldn't TypeScript and Elm completely eliminate this confusion, since it shows you what you can pass in?

Reply:
If I were to vote, I'd just use the unabbreviated words as much as possible (this is specifically in regards to abbreviated vs unabbreviated, not `textColor` vs `foregroundColor`). I don't think there's a big enough advantage to abbreviating things. Saves a tiny about of typing time and yeah a little easier to ready once you're experienced, but then forces you to have remember whether something was abbreviated or not and it's extra friction for. Clearly we won't be abbreviating everything, "CCIIHSTODST" (see if you can work out that one), so it's not super clear what the rules are for when to abbreviate and when not to. `fgColor` seems like a half measure to me!

Also, what's obvious to you might not be obvious to someone else. Let me recount my recent experience when I honestly thought `md` stood for "middle" and it was to do with responsitivity, not "markdown" (there was a CSS class something like `mdHeading`. I wasted at least an hour because of this thinking that we had a responsive system in place for Kaizen Site.

Reply:
Totally agree with you @mbylstra, but in this particular case `backgroundColor` has this problem:

> Following this will result in name collisions with CSS properties (background-color is backgroundColor when represented as an object in React).

Do you know how we could get around that here?

Reply:
It's not a _technical_ limitation though right? One is a css property, and one is a prop, so they wouldn't technically clash. It's slightly confusing, but I think the type system should make it pretty clear. It is a bit of a tricky one, but I'd still lean towards no abbreviations.

Reply:
Also, I'm assuming that if we reject `textColor` and `backgroundColor` on the basis that one of them conflicts with a CSS property, then we'll make this a rule going forward and consistently reject every prop name that conflicts with a CSS property, yeah?

Reply:
@mbylstra I agree, TS and Elm will clear up the confusion.

Reply:
But I'm happy to step back and trust you to make a decision on this one (this kind of thing can easily go around and circles!)

Reply:
This is where it gets complicated. Is it my decision to make because I started the PR? If @ActuallyACat disagrees, should I trust her to make the decision, since she mentioned she has strong opinions about how our component APIs should be structured? I feel pressure not to drag the conversation on, since this is a thing we've been doing a lot recently, but I also feel like it'd be irresponsible of me to not mention to you two when I spot risks.

Reply:
Hehe, yeah it is tricky. What I really meant to say is that my comments are my 2c, but I don't feel super strongly about it, and it's harder with 3 people, so I'm stepping out so that at least there's only two people that have to agree, which should be easier than 3!

Reply:
Alright. I'm going to ask that @ActuallyACat has the final say on this, and I'll lay out the arguments against her idea for her to consider before she makes the call:

• `fg` and `bg` could be unintuitive and confusing for first-time users, particularly if there is no `bg` and they just see `fg` on its own and don't know what it is. Worst-case scenario could be that someone uses the Heading or Paragraph component, wants to change the colour but doesn't realise that `fg` does this, but they don't ask us, and instead create a container div that reaches in and changes the colour from outside.
• `fg` and `bg` are not part of any system that our designers use to talk about colour, so this arguably goes against our efforts to make the design system feel as one (subcategory: "designers and engineers speak the same language").
• even though `backgroundColor` looks like a CSS property, TypeScript and Elm would prevent anyone from passing in hex values or the like, and would likely clear up confusion about what they can pass in. `textColor` isn't a CSS property so it doesn't have the issue at all. Both of these terms match the way that designers talk about colour and are intuitive to read at first glance.

As long as I've clearly articulated all those, I'm happy to go with whatever @ActuallyACat chooses since I'm happy she has all the info that I could have provided. 👍

Reply:
These are all 💯:
`fgColor` and `bgColor`
`foregroundColor` and `backgroundColor`
`fg` and `bg`

Speaking generally, I think overloading language should be something we're mindful of and call out when we see it. It's all about tradeoffs, after all. You're both right that TS types should make this easy to reason about in this case, so this trade-off seems like a reasonable one. 

`textColor` particularly irks me because it's a distinction we shouldn't need to make (referencing my composition example).

Reply:
To summarise the trade-offs:

- **designers and developers speak the same language** (`textColor`, `color` ✅ `foreground`, `fg`, `fgColor` ❌)
- **props don’t look like CSS properties and thus potentially overload the terms** (`textColor`, `foreground`, `fg`, `fgColor` ✅, `color`, `backgroundColor` ❌)
- **having unnecessary words in prop names is unnecessary, and would mean we’d have to make further unnecessary distictions later** (`color` ✅ , `textColor` ❌)

After chatting to @ActuallyACat about this, I'm going to go with `backgroundColor` and `color` since they seem to both work for 2 out of 3 trade-offs and the one they don't satisfy isn't that bad we think.

Reply:
One thing I just want a little clarity on: will the heading and paragraph components take `backgroundColor` props (regardless of whether it's named `bg`, `bgColor`, etc), and if so, why?

Reply:
Nice summary, thanks @sebpearce 

@mbylstra I don't think there's a compelling use case for that, no. But these colour props I would consider to be "common props" (like `tag`) that will most likely be used across other components. I could see these being used on icons, cards, and other more atomic components.